File: src/main/java/com/rednorte/sigle/citas_service/controller/CitaController.java
package com.rednorte.sigle.citas_service.controller;

import com.rednorte.sigle.citas_service.model.Cancelacion;
import com.rednorte.sigle.citas_service.model.CanceladoPor;
import com.rednorte.sigle.citas_service.model.Cita;
import com.rednorte.sigle.citas_service.service.CitaService;

import jakarta.validation.Valid;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import lombok.Data;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.time.LocalDate;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/citas")
@RequiredArgsConstructor
public class CitaController {

    private final CitaService citaService;

    @GetMapping
    public ResponseEntity<List<Cita>> getAll() {
        return ResponseEntity.ok(citaService.getAll());
    }

    @GetMapping("/{id}")
    public ResponseEntity<Cita> getById(@PathVariable Long id) {
        return ResponseEntity.ok(citaService.getById(id));
    }

    @GetMapping("/paciente/{pacienteId}")
    public ResponseEntity<List<Cita>> getByPacienteId(@PathVariable Long pacienteId) {
        return ResponseEntity.ok(citaService.getByPacienteId(pacienteId));
    }

    @GetMapping("/paciente/{pacienteId}/paginado")
    public ResponseEntity<Map<String, Object>> getByPacienteIdPaginado(
            @PathVariable Long pacienteId,
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "10") int size) {
        Page<Cita> resultado = citaService.getByPacienteIdPaginado(pacienteId, PageRequest.of(page, size));
        Map<String, Object> response = new HashMap<>();
        response.put("content", resultado.getContent());
        response.put("totalPages", resultado.getTotalPages());
        response.put("totalElements", resultado.getTotalElements());
        response.put("currentPage", resultado.getNumber());
        return ResponseEntity.ok(response);
    }

    @GetMapping("/medico/{medicoId}")
    public ResponseEntity<List<Cita>> getByMedicoId(@PathVariable Long medicoId) {
        return ResponseEntity.ok(citaService.getByMedicoId(medicoId));
    }

    @GetMapping("/medico/{medicoId}/horas-ocupadas")
    public ResponseEntity<List<String>> getHorasOcupadas(
            @PathVariable Long medicoId,
            @RequestParam String fecha) {
        return ResponseEntity.ok(citaService.getHorasOcupadas(medicoId, LocalDate.parse(fecha)));
    }

    @PostMapping("/agendar")
    public ResponseEntity<Cita> agendarCita(@Valid @RequestBody AgendarRequest r) {
        return ResponseEntity.ok(citaService.agendarCita(r.getCita(), r.getMedicoId()));
    }

    @PutMapping("/{id}")
    public ResponseEntity<Cita> update(@PathVariable Long id, @Valid @RequestBody AgendarRequest r) {
        return ResponseEntity.ok(citaService.update(id, r.getCita(), r.getMedicoId()));
    }

    @PostMapping("/{id}/cancelar")
    public ResponseEntity<Cancelacion> cancelarCita(@PathVariable Long id, @Valid @RequestBody CancelacionRequest request) {
        Cancelacion cancelacion = citaService.cancelarCita(id, request.getMotivo(), request.getCanceladoPor());
        return ResponseEntity.ok(cancelacion);
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<Void> delete(@PathVariable Long id) {
        citaService.delete(id);
        return ResponseEntity.ok().build();
    }

    @Data
    public static class CancelacionRequest {
        @NotBlank(message = "El motivo es obligatorio")
        private String motivo;

        @NotNull(message = "Debe indicar quién cancela")
        private CanceladoPor canceladoPor;
    }

    @Data
    public static class AgendarRequest {
        @NotNull(message = "La cita es obligatoria")
        private Cita cita;

        @NotNull(message = "El médico es obligatorio")
        private Long medicoId;
    }
}

File: src/controllers/citaController.js
const service = require('../services/citaService');

const getAll = async (req, res, next) => {
  try { res.json(await service.getAll()); } catch (e) { next(e); }
};

const getById = async (req, res, next) => {
  try { res.json(await service.getById(req.params.id)); } catch (e) { next(e); }
};

const getByPacienteId = async (req, res, next) => {
  try { res.json(await service.getByPacienteId(req.params.pacienteId)); } catch (e) { next(e); }
};

const getByPacienteIdPaginado = async (req, res, next) => {
  try {
    const page = parseInt(req.query.page) || 0;
    const size = parseInt(req.query.size) || 10;
    res.json(await service.getByPacienteIdPaginado(req.params.pacienteId, page, size));
  } catch (e) { next(e); }
};

const getByMedicoId = async (req, res, next) => {
  try { res.json(await service.getByMedicoId(req.params.medicoId)); } catch (e) { next(e); }
};

const getHorasOcupadas = async (req, res, next) => {
  try {
    const { fecha } = req.query;
    if (!fecha) return res.status(400).json({ error: 'El parámetro fecha es obligatorio (YYYY-MM-DD)' });
    res.json(await service.getHorasOcupadas(req.params.medicoId, fecha));
  } catch (e) { next(e); }
};

const agendarCita = async (req, res, next) => {
  try {
    const { cita, medicoId } = req.body;
    if (!cita || !medicoId) return res.status(400).json({ error: 'cita y medicoId son obligatorios' });
    res.json(await service.agendarCita(cita, medicoId));
  } catch (e) { next(e); }
};

const update = async (req, res, next) => {
  try {
    const { cita, medicoId } = req.body;
    res.json(await service.update(req.params.id, cita || req.body, medicoId));
  } catch (e) { next(e); }
};

const cancelarCita = async (req, res, next) => {
  try {
    const { motivo, canceladoPor } = req.body;
    res.json(await service.cancelarCita(req.params.id, motivo, canceladoPor));
  } catch (e) { next(e); }
};

const remove = async (req, res, next) => {
  try { await service.remove(req.params.id); res.json({ message: 'Cita eliminada' }); } catch (e) { next(e); }
};

module.exports = {
  getAll, getById, getByPacienteId, getByPacienteIdPaginado,
  getByMedicoId, getHorasOcupadas, agendarCita, update, cancelarCita, remove,
};
